### PR TITLE
DRILL-3535: Add support for Drop Table

### PR DESCRIPTION
**Pull request by mehant**

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillPathFilter.java [hunk 1]
```diff
@@ -23,7 +23,7 @@
 public class DrillPathFilter extends Utils.OutputFileUtils.OutputFilesFilter {
   @Override
   public boolean accept(Path path) {
-    if (path.toString().contains("_metadata")) {
+    if (path.getName().startsWith(DrillFileSystem.HIDDEN_FILE_PREFIX)) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillPathFilter.java]: We should add dot prefixed files/folders here.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java [hunk 2]
```diff
@@ -72,7 +72,7 @@ public boolean supportDirectoryReads() {
 
   @Override
   public FormatSelection isReadable(DrillFileSystem fs, FileSelection selection) throws IOException {
-    if (isReadable(fs, selection.getFirstPath(fs))) {
+    if (isFileReadable(fs, selection.getFirstPath(fs))) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java]: I think we need to check all of the selection for Drop, not just the first.  Otherwise we risk blowing away a directory which has a bunch of files, where only the first is readable.  Right?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 3]
```diff
@@ -321,8 +327,101 @@ public DrillTable create(String key) {
...
+
     @Override
     public void destroy(DrillTable value) {
     }
+
+    /**
+     * Check if the table contains homogenenous files that can be read by Drill. Eg: parquet, json csv etc.
+     * However if it contains more than one of these formats or a totally different file format that Drill cannot
+     * understand then we will raise an exception.
+     * @param key
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Key?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 4]
```diff
@@ -321,8 +327,101 @@ public DrillTable create(String key) {
...
+     * @return
+     * @throws IOException
+     */
+    private boolean isHomogeneous(String key) throws IOException {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Why would we only check isHomogeneous for drop and not for query?  Thoughts here?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java (hunk 2 above)
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java]: Why did you change the name here?  If I recall correctly, the Parquet format matcher doesn't necessarily need to read a file and can match at the directory level.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 5]
```diff
@@ -104,6 +107,9 @@ public WorkspaceSchemaFactory(DrillConfig drillConfig, FileSystemPlugin plugin,
       final FormatMatcher fallbackMatcher = new BasicFormatMatcher(formatPlugin,
           ImmutableList.of(Pattern.compile(".*")), ImmutableList.<MagicString>of());
       fileMatchers.add(fallbackMatcher);
+      dropFileMatchers = fileMatchers.subList(0, fileMatchers.size() - 1);
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Why would we have different matchers for drop file.  Shouldn't we simply share the behavior across all matching behaviors?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 6]
```diff
@@ -321,8 +327,101 @@ public DrillTable create(String key) {
...
+      String defaultLocation = getDefaultLocation();
+      try {
+        if (!isHomogeneous(table)) {
+          throw UserException
+              .validationError()
+              .message("Table contains different file formats. \n" +
+                  "Drop Table is only supported for directories that contain homogeneous file formats consumable by Drill")
+              .build(logger);
+        }
+        fs.rename(new Path(defaultLocation, table), new Path(defaultLocation, dirName));
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: If this rename is to provide transactional capabilities, shouldn't it also be uniquely named (rather than simply underscore prefixed)?

Shouldn't we capture a failure of rename and return the user that the table no longer exists? (in the case of a race)

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 7]
```diff
@@ -321,8 +327,101 @@ public DrillTable create(String key) {
...
+        fs.delete(new Path(defaultLocation, dirName), true);
+      } catch (IOException e) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Shouldn't the permission error only be constrained to certain types of IOExceptions, not all?  It seems like failing on the delete would indicate something else.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java (hunk 5 above)
Comment (mehant) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: The matchers are the same except for the pattern associated with defaultInputFormat. Because the pattern associated with defaultInputFormat is .* and we don't want to match and delete on this.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java (hunk 4 above)
Comment (mehant) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: The only reason was to avoid the performance penalty for read that would be ensued by these checks and simply going ahead optimistically and failing later if we hit different formats.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java (hunk 6 above)
Comment (mehant) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: I will add a unique identifier in addition to the underscore prefix.

I modified the exception handling below to separate out AccessControlException and the generic IOException to handle permission errors and other errors respectively. In the case of a race with concurrent renames the file system would throw a FileNotFoundException which would be caught by the IOException and reported to the user.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java (hunk 2 above)
Comment (mehant) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/BasicFormatMatcher.java]: You are right, ParquetFormatMatcher overrides isReadable to first check if the directory contains the metadata file, if not it delegates to the base class method to check if a particular file is readable or not. 

The existing naming is a bit confusing. In BasicFormatMatcher there are two methods with the same name:

1. FormatSelection isReadable(DrillFileSystem fs, FileSelection selection) -> operates on directory level
2. boolean isReadable(DrillFileSystem fs, FileStatus status) -> operates on a single file level.

I've renamed the second to be called isFileReadable and exposed it in the abstract class so drop can invoke it on every file.

**Comment by mehant** [comment]
@jacques-n uploaded patch with your review comments. Please take a look.

**Comment by jacques-n** [comment]
LGTM +1
